compiler/ballerina-lang/src/main/java/io/ballerina/projects/JBallerinaBackend.java

```diff
@@ -177,8 +177,10 @@ private void performCodeGen() {
                         new PackageDiagnostic(diagnostic, moduleContext.descriptor(), moduleContext.project()));
             }
 
-            //TODO: remove this once ballerina-lang#41407 is fixed
             ModuleContext.shrinkDocuments(moduleContext);
+            if (moduleContext.project().kind() == ProjectKind.BALA_PROJECT) {
```

Comment:
Why do we need this check?

Reply:
We only make `null` the BLangPackages of the dependencies of a Ballerina package (e.g. the imported modules of this Ballerina snippet https://github.com/ballerina-platform/ballerina-lang/issues/41407#issue-1902210827 here) because the `BLangPackage`s of the Ballerina package is needed for other tasks such as to return the semantic model to be used by the language server and vscode. The `if` check returns true for such modules.